sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -650,7 +650,9 @@ object ColumnPruning extends Rule[LogicalPlan] {
    */
   private def removeProjectBeforeFilter(plan: LogicalPlan): LogicalPlan = plan transformUp {
     case p1 @ Project(_, f @ Filter(_, p2 @ Project(_, child)))
-      if p2.outputSet.subsetOf(child.outputSet) =>
+      if p2.outputSet.subsetOf(child.outputSet) &&
+        // We only remove attribute-only project.
+        p2.projectList.forall(_.isInstanceOf[AttributeReference]) =>
```

Comment:
How can we remove project that's not attribute-only?

Reply:
I'd say it was wrong previously, but if a project's output has same expr IDs with its child, it's usually attribute-only.

Reply:
Mmmmh... I may be missing something, but I'd imagine a case like this:
```
select a, b from
(select a, b, very_expensive_operation as c from ... where a = 1)
```
Before this change, would be optimized as:
```
select a, b from
(select a, b from ... where a = 1)
```
while after it is not. Am I wrong?

Reply:
In above case, it has a Alias in project list, so it's not an attribute-only project. And I think it also create new attr c, so `p2.outputSet.subsetOf(child.outputSet)` is not met too.

Reply:
I see now, sorry. Why do we need this? Seems an unrelated change to the fix in this PR, isn't it?

Reply:
oh, the issue was seen in previous comment https://github.com/apache/spark/pull/25204/commits/33441a32412eacbe84f2b188f046b2959a8b60ee. It was overwritten now.

We added a column for count bug. The column checks a always-true leading column `alwaysTrueExpr`, returns special value if `alwaysTrueExpr` is null, to simulate empty input case.

This column reuses expr id of original output in the subquery. In non-foldable expression case, the added column in a potential Project-Filter-Project, will be trimmed by `removeProjectBeforeFilter`, because the second project meets `p2.outputSet.subsetOf(child.outputSet)`.

My original fix is to create an expr id. Replace original expr id with new one in the subquery. Looks complicated. This seems a simple fix, and looks reasonable.